core/Pos/Core/Types.hs

```diff
@@ -120,29 +123,78 @@ type AddressHash = AbstractHash Blake2b_224
 -- | Stakeholder identifier (stakeholders are identified by their public keys)
 type StakeholderId = AddressHash PublicKey
 
--- | Address is where you can send coins.
-data Address
-    = PubKeyAddress
-          { addrKeyHash      :: !(AddressHash PublicKey)
-          , addrPkAttributes :: !(Attributes AddrPkAttrs) }
-    | ScriptAddress
-          { addrScriptHash :: !(AddressHash Script) }
-    | RedeemAddress
-          { addrRedeemKeyHash :: !(AddressHash RedeemPublicKey) }
-    | UnknownAddressType !Word8 !ByteString
-    deriving (Eq, Ord, Generic, Typeable, Show)
-
-instance NFData Address
+-- | A mapping between stakeholders and they stakes.
+type StakesMap = HashMap StakeholderId Coin
 
-newtype AddrPkAttrs = AddrPkAttrs
-    { addrPkDerivationPath :: Maybe HDAddressPayload
-    } deriving (Eq, Ord, Show, Generic, Typeable, NFData)
+-- | Data which is bound to an address and must be revealed in order
+-- to spend coins belonging to this address.
+data AddrSpendingData
+    = PubKeyASD !PublicKey
+    -- ^ Funds can be spent by revealing a 'PublicKey' and providing a
+    -- valid signature.
+    | ScriptASD !Script
+    -- ^ Funds can be spent by revealing a 'Script' and providing a
+    -- redeemer 'Script'.
+    | RedeemASD !RedeemPublicKey
+    -- ^ Funds can be spent by revealing a 'RedeemScript' and providing a
+    -- valid signature.
+    | UnknownASD !Word8 !ByteString
+    -- ^ Unknown type of spending data. It consists of a tag and
+    -- arbitrary 'ByteString'. It allows us to introduce a new type of
+    -- spending data via softfork.
+    deriving (Eq, Generic, Typeable, Show)
+
+-- | Type of an address. It corresponds to constructors of
+-- 'AddrSpendingData'. It's separated, because 'Address' doesn't store
+-- 'AddrSpendingData', but we want to know its type.
+data AddrType
+    = ATPubKey
+    | ATScript
+    | ATRedeem
+    | ATUnknown !Word8
+    deriving (Eq, Ord, Generic, Typeable, Show)
 
-instance Default AddrPkAttrs where
-    def = AddrPkAttrs Nothing
+-- TODO [CSL-1489] This stake distribution is currently unused!
+
+-- | Stake distribution associated with an address.
+data AddrStakeDistribution
+    = BootstrapEraDistr
+    -- ^ Stake distribution for bootstrap era.
+    | SingleKeyDistr !StakeholderId
+    -- ^ Stake distribution stating that all stake should go to the given stakeholder.
+    | MultiKeyDistr ![(StakeholderId, Coin)]
```

Comment:
Because initially I planned to preserve old logic such that empty means _stake should go to stakeholder from an address implicitly_, but then I realized it was impossible.
Good point, I will use `NonEmpty` in CSL-1489 (currently this distr is ignored anyway, so doesn't matter).